sklearn/metrics/_ranking.py

```diff
@@ -1419,3 +1419,120 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
     gain = _ndcg_sample_scores(y_true, y_score, k=k, ignore_ties=ignore_ties)
     return np.average(gain, weights=sample_weight)
+
+
+def top_k_accuracy_score(y_true, y_score, k=3, normalize=True,
+                         sample_weight=None):
+    """Top-k Accuracy multiclass classification score.
+
+    This metric computes the number of times where the correct label is among
+    the top `k` labels predicted (ranked by predicted scores). Note that the
+    multilabel case isn't covered here.
+
+    Read more in the :ref:`User Guide <top_k_accuracy_score>`
+
+    Parameters
+    ----------
+    y_true : array-like of shape (n_samples,)
+        True labels. Expected to be a sequence of int in the range
+        [0, n_classes - 1].
+
+    y_score : array-like of shape (n_samples, n_classes)
+        Target scores. These can be either probability estimates or
+        non-thresholded decision values as returned by :term:`predict_proba` or
+        :term:`decision_function`, respectively.
+
+    k : int, default=3
+        Number of guesses allowed to find the correct label.
+
+    normalize : bool, default=True
+        If `True`, return the fraction of correctly classified samples.
+        Otherwise, return the number of correctly classified samples.
+
+    sample_weight : array-like of shape (n_samples,), default=None
+        Sample weights. If `None`, all samples are given the same weight.
+
+    Returns
+    -------
+    score : float
+        The top-k accuracy score. The best performance is 1 with
+        `normalize == True` and the number of samples with
+        `normalize == False`.
+
+    See also
+    --------
+    accuracy_score
+
+    Notes
+    -----
+    In cases where two or more labels are assigned equal predicted scores,
+    the labels with the smallest indices will be chosen first. This might
+    impact the result if the correct label falls after the threshold because
+    of that.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn.metrics import top_k_accuracy_score
+    >>> y_true = np.array([0, 1, 2, 2])
+    >>> y_score = np.array([[0.5, 0.2, 0.2],  # 0 is in top 2
+    ...                     [0.3, 0.4, 0.2],  # 1 is in top 2
+    ...                     [0.2, 0.4, 0.3],  # 2 is in top 2
+    ...                     [0.7, 0.2, 0.1]]) # 2 isn't in top 2
+    >>> top_k_accuracy_score(y_true, y_score, k=2)
+    0.75
+    >>> top_k_accuracy_score(y_true, y_score, k=2, normalize=False)
+    3
+    >>> top_k_accuracy_score([0, 0, 0, 0], y_score, k=2)
+    0.75
+
+    """
+    if k == 1:
+        raise ValueError(
+            "'k'=1 is equivalent to 'metrics.accuracy_score'. Please, use "
+            "that function instead."
+        )
+
+    check_consistent_length(y_true, y_score, sample_weight)
+    y_true_type = type_of_target(y_true)
+
+    if y_true_type not in {'binary', 'multiclass'}:
+        raise ValueError(
+            "'y_true' type must be 'binary' or 'multiclass', got "
```

Comment:
shouldn't accept `binary` if we're not accepting `n_classes=2`.

Reply:
`y_true` could contains a subset of the classes, e.g. `[0, 0, 1]`

Reply:
but then we'll raise because the number of classes in `y_true` is not the same as the number of columns in `y_pred`.

Reply:
As we assume that `y_true` doesn't have to contain each possible label we are not checking that it has the same number of classes than `y_pred` number of columns. Instead we check that inferred `y_true` number of classes is less or equal to `y_score.shape[1]`